HDDS-3725. Ozone sh volume client support quota option.

captainzmc opened the pull request:
## What changes were proposed in this pull request?
Currently, Ozone's support for Quota is incomplete, and currently only supports setting volume capacity Quota by command. The Quota setting based on quantity is not yet supported. The purpose of this PR is to support this.
In addition, the current Quota setting does not take effect. HDDS-541 gives all the work needed to perfect Quota.
This PR is a subtask of HDDS-541. Design documentation for Ozone Quota:
https://docs.google.com/document/d/1ohbGn5N7FN6OD15xMShHH2SrtZRYx0-zUf9vjatn_OM/edit

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3725

## How was this patch tested?

Added UT.

cxorm commented on hadoop-hdds/docs/content/shell/VolumeCommands.md:
```diff
@@ -37,12 +37,13 @@ assign it to a user.
-| -u, \-\-user                     |  Required, The name of the user who owns this volume. This user can create, buckets and keys on this volume.                                       |
+| -sq, \-\-spaceQuota             | Optional, This argument that specifies the maximum size this volume can use in the Ozone cluster.                                        |
+| -q, \-\-quota                  | Optional, This argument that specifies the number of bucket in this volume can use in the Ozone cluster.                                        |
+| -u, \-\-user                   | Required, The name of the user who owns this volume. This user can create, buckets and keys on this volume.                                       |
```
```suggestion
| -u, \-\-user                   | Required, The name of the user who owns this volume. This user can create buckets and keys on this volume.                                       |
```

cxorm commented on hadoop-hdds/docs/content/shell/VolumeCommands.md:
```diff
@@ -37,12 +37,13 @@ assign it to a user.
-| -q, \-\-quota                    | Optional, This argument that specifies the maximum size this volume can use in the Ozone cluster.                    |
-| -u, \-\-user                     |  Required, The name of the user who owns this volume. This user can create, buckets and keys on this volume.                                       |
+| -sq, \-\-spaceQuota             | Optional, This argument that specifies the maximum size this volume can use in the Ozone cluster.                                        |
+| -q, \-\-quota                  | Optional, This argument that specifies the number of bucket in this volume can use in the Ozone cluster.                                        |
```
```suggestion
| -q, \-\-quota                  | Optional, This argument that specifies the number of buckets in this volume can use in the Ozone cluster.                                        |
```

cxorm commented on hadoop-hdds/docs/content/shell/VolumeCommands.md:
```diff
@@ -103,12 +104,13 @@ The volume update command allows changing of owner and quota on a given volume.
-| -q, \-\-quota                    | Optional, This argument that specifies the maximum size this volume can use in the Ozone cluster.                    |
-| -u, \-\-user                     |  Optional, The name of the user who owns this volume. This user can create, buckets and keys on this volume.                                       |
+| -sq, \-\-spaceQuota             | Optional, This argument that specifies the maximum size this volume can use in the Ozone cluster.                                        |
+| -q, \-\-quota                  | Optional, This argument that specifies the number of bucket in this volume can use in the Ozone cluster.                                        |
```
```suggestion
| -q, \-\-quota                  | Optional, This argument that specifies the number of buckets in this volume can use in the Ozone cluster.                                        |
```

cxorm commented on hadoop-hdds/docs/content/shell/VolumeCommands.md:
```diff
@@ -103,12 +104,13 @@ The volume update command allows changing of owner and quota on a given volume.
-| -u, \-\-user                     |  Optional, The name of the user who owns this volume. This user can create, buckets and keys on this volume.                                       |
+| -sq, \-\-spaceQuota             | Optional, This argument that specifies the maximum size this volume can use in the Ozone cluster.                                        |
+| -q, \-\-quota                  | Optional, This argument that specifies the number of bucket in this volume can use in the Ozone cluster.                                        |
+| -u, \-\-user                   | Optional, The name of the user who owns this volume. This user can create, buckets and keys on this volume.                                       |
```
```suggestion
| -u, \-\-user                   | Optional, The name of the user who owns this volume. This user can create buckets and keys on this volume.                                       |
```

cxorm commented on hadoop-hdds/interface-client/src/main/proto/proto.lock:
```diff
@@ -1935,4 +1935,4 @@
     }
   ]
-}
+}
```
Seems we could not modify here : )

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java:
```diff
@@ -61,9 +61,13 @@
   private String owner;
   /**
-   * Quota allocated for the Volume.
+   * Bytes Quota allocated for the Volume.
```
Suggestion :
"Quota of bytes allocated for the Volume."

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java:
```diff
@@ -61,9 +61,13 @@
-  private long quotaInBytes;
+  private long storagespaceQuota;
+  /**
+   * Bucket Counts Quota allocated for the Volume.
```
Suggestion :
"Quota of bucket count allocated for the Volume."

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java:
```diff
@@ -248,12 +267,14 @@ public boolean setOwner(String userName) throws IOException {
-   * @param quota new quota
-   * @throws IOException
+   *
+   * @param quota@throws IOException
```
Suggestion :
Separate the annotations and add the comments.

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java:
```diff
@@ -204,10 +215,18 @@ public String getOwner() {
   }
 
+  /**
+   * Returns Bucket Quota count allocated for the Volume.
```
Suggestion :
"Returns quota of bucket count allocated for the Volume."

cxorm commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/CreateVolumeHandler.java:
```diff
@@ -40,10 +41,15 @@
 
+  @Option(names = {"--spaceQuota", "-sq"},
+      description =
+          "Quota Bytes of the newly created volume (eg. 1GB)")
```
```suggestion
          "Quota in bytes of the newly created volume (eg. 1GB)")
```

cxorm commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/CreateVolumeHandler.java:
```diff
@@ -40,10 +41,15 @@
       description =
-          "Quota of the newly created volume (eg. 1G)")
-  private String quota;
+          "Quota count of the newly created volume (eg. 5)")
```
Correct me if I miss something.
Would it be better to update the description to "Bucket counts of the newly created volume (eg. 5) " ?

cxorm commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/CreateVolumeHandler.java:
```diff
@@ -40,10 +41,15 @@
       description = "Owner of of the volume")
```
```suggestion
      description = "Owner of the volume")
```

cxorm commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/CreateVolumeHandler.java:
```diff
@@ -40,10 +41,15 @@
       description = "Owner of of the volume")
   private String ownerName;
 
+  @Option(names = {"--spaceQuota", "-sq"},
```
Do you mind change the "-sq" to "-s" (or other alphabet) ?

I think the command-line promotion is better if it is alignment. 
(And take this way, it would be left-alignment)

![out of alignment](https://i.imgur.com/nXhbsfW.png)
![left alignment](https://i.imgur.com/xzUAiZe.png)

captainzmc commented:
Thanks @cxorm for your review. I have fixed all the problems mentioned above. The command line display has also changed(As shows below).  Could you please take another look？ 
![image](https://user-images.githubusercontent.com/13825159/88183192-ba70ba80-cc63-11ea-9ff3-0e1c8b7b0717.png)

cxorm commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1706,21 +1706,24 @@ public boolean setOwner(String volume, String owner) throws IOException {
   @Override
-  public void setQuota(String volume, long quota) throws IOException {
-    if (isAclEnabled) {
+  public void setQuota(String volume, long namespaceQuota,
```
Instead of `OzoneManager.java`, this request is processed in `OMVolumeSetQuotaRequest.java` .

We could not update code-snippet here.

cxorm commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/VolumeManager.java:
```diff
@@ -49,10 +49,12 @@ void setOwner(String volume, String owner)
    * @throws IOException
    */
-  void setQuota(String volume, long quota) throws IOException;
+  void setQuota(String volume, long namespaceQuota,
```
This part is the same as comments on `OzoneManager.java `

cxorm replied:
This part is the same as comments on `OzoneManager.java`

cxorm commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/VolumeManagerImpl.java:
```diff
@@ -273,12 +272,14 @@ private void setOwnerCommitToDB(UserVolumeInfo oldOwnerVolumeList,
    */
   @Override
-  public void setQuota(String volume, long quota) throws IOException {
+  public void setQuota(String volume, long namespaceQuota,
+                       long storagespaceQuota) throws IOException {
```
This part is the same as comments on OzoneManager.java

cxorm commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/CreateVolumeHandler.java:
```diff
@@ -58,9 +62,12 @@ protected void execute(OzoneClient client, OzoneAddress address)
         .setOwner(ownerName);
-    if (quota != null) {
-      volumeArgsBuilder.setQuota(quota);
+    if (storagespaceQuota!= null) {
```
```suggestion
    if (storagespaceQuota != null) {
```

cxorm replied:
```suggestion
    if (storagespaceQuota != null) {
```

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/VolumeArgs.java:
```diff
@@ -122,8 +136,13 @@ public String getQuota() {
+      return this;
+    }
+
+    public VolumeArgs.Builder setNamespaceQuotaQuota(long quota) {
```
```suggestion
    public VolumeArgs.Builder setNamespaceQuota(long quota) {
```

cxorm replied:
```suggestion
    public VolumeArgs.Builder setNamespaceQuota(long quota) {
```

cxorm commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/CreateVolumeHandler.java:
```diff
@@ -58,9 +62,12 @@ protected void execute(OzoneClient client, OzoneAddress address)
+      volumeArgsBuilder.setStoragespaceQuota(storagespaceQuota);
     }
+
+    volumeArgsBuilder.setNamespaceQuotaQuota(namespaceQuota);
```
The function name seems a little redundant.

Suggestion is commented on `VolumeArgs.java`

cxorm replied:
The function name seems a little redundant.

Suggestion is commented on `VolumeArgs.java`.

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/VolumeArgs.java:
```diff
@@ -33,25 +33,30 @@
    * @param admin Administrator's name.
    * @param owner Volume owner's name
-   * @param quota Volume Quota.
+   * @param storagespaceQuota Volume Quota.
+   * @param namespaceQuota Volume Quota.
```
I think we could update the description of the parameter consistent with shell-command IMHO.

cxorm replied:
I think we could update the description of the parameter consistent with shell-command IMHO.

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/VolumeArgs.java:
```diff
@@ -73,11 +78,19 @@ public String getOwner() {
+  }
+
+  /**
+   * Returns Volume Quota in counts.
```
```suggestion
   * Returns Volume Quota in bucket counts.
```

cxorm replied:
```suggestion
   * Returns Volume Quota in bucket counts.
```

cxorm commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/StringUtils.java:
```diff
@@ -149,6 +150,13 @@ public static String createStartupShutdownMessage(VersionInfo versionInfo,
         "  java = " + System.getProperty("java.version"));
   }
 
+  /**
+   * The same as String.format(Locale.ENGLISH, format, objects).
+   */
+  public static String format(final String format, final Object... objects) {
+    return String.format(Locale.ENGLISH, format, objects);
+  }
+
```
I didn't found usage of this method in all places.
Could you tell me why we add this part if I miss something ?

cxorm replied:
I didn't found usage of this method in all places.
Could you tell me why we add this part if I miss something ?

cxorm commented:
Thanks @captainzmc for the work.
Overall the functionalily-part is good and some suggestions is added inline, test-related-part would be reviewd soon.

Here I have two Questions IMHO.

(1) Could we let the variable-pair be consistent ?
I think it would better to use one of (`storagespaceQuota`/ `namespaceQuota`) and (`quotaInBytes`/ `quotaInCounts`).
If it have other purpose to use both, please let me know.

(2) Would we add limitation of creating volume and bucket in later patch ?
I deployed this patch on my machine, and created volume without setting both of quota.
I found that the storage of volume exceeds my local-storage.
And I found we could create bucket even if the quota of bucket is -1.
![create_bucket](https://i.imgur.com/oN1Jf4X.png)

cxorm commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/UpdateVolumeHandler.java:
```diff
@@ -39,22 +40,34 @@
+      countQuota = namespaceQuota;
     }
 
+    volume.setQuota(
+        OzoneQuota.getOzoneQuota(spaceQuota, countQuota));
```
We could not violate `checkstyle` if we use one line here.

cxorm replied:
We could not violate `checkstyle` if we use one line here.

cxorm commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -25,26 +25,75 @@
   /** Quota Units.*/
   public enum Units {UNDEFINED, BYTES, KB, MB, GB, TB}
 
+  // Quota to decide how many buckets or keys can be created.
```
```suggestion
  // Quota to decide how many buckets can be created.
```

cxorm replied:
```suggestion
  // Quota to decide how many buckets can be created.
```

cxorm commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -25,26 +25,75 @@
+  private long storagespaceQuota;
+  private RawStorageSpaceQuota rawStoragespaceQuota;
+
+  private static class RawStorageSpaceQuota {
```
Here we could add comment to the class `RawStorageSpaceQuota` IMHO.

cxorm replied:
The RawStorageSpace is a little confuse. Here we could add comment addressed to the class `RawStorageSpaceQuota` IMHO.

cxorm commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -53,26 +102,28 @@ public long getSize() {
-  public OzoneQuota(long size, Units unit) {
-    this.size = size;
-    this.unit = unit;
+  private OzoneQuota(long namespaceQuota,
+                     RawStorageSpaceQuota rawStoragespaceQuota) {
```
```suggestion
  private OzoneQuota(long namespaceQuota,
      RawStorageSpaceQuota rawStoragespaceQuota) {
```

cxorm replied:
```suggestion
  private OzoneQuota(long namespaceQuota,
      RawStorageSpaceQuota rawStoragespaceQuota) {
```

cxorm commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -82,25 +133,28 @@ public OzoneQuota(long size, Units unit) {
    * @return OzoneQuota object
    */
-  public static OzoneQuota parseQuota(String quotaString) {
+  public static OzoneQuota parseQuota(String storagespaceQuotaStr,
+                                      long namespaceQuota) {
```
```suggestion
  public static OzoneQuota parseQuota(String storagespaceQuotaStr,
      long namespaceQuota) {
```

cxorm replied:
```suggestion
  public static OzoneQuota parseQuota(String storagespaceQuotaStr,
      long namespaceQuota) {
```

cxorm commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java:
```diff
@@ -204,6 +204,16 @@ public static Versioning getVersioning(boolean versioning) {
    */
   public static final long MAX_QUOTA_IN_BYTES = 1024L * 1024 * TB;
 
+  /**
+   * Quota value.
+   */
+  public static final long QUOTA_COUNT_RESET = -1;
```
Could you please update the comment to "Quota of bucket counts" and add the description of meaning of "-1" here ?

cxorm replied:
Could you please update the comment to "Quota of bucket counts" and add the description of meaning of "-1" here ?

cxorm commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -143,57 +204,53 @@ public static OzoneQuota parseQuota(String quotaString) {
    * @return OzoneQuota object
    */
-  public static OzoneQuota getOzoneQuota(long sizeInBytes) {
+  public static OzoneQuota getOzoneQuota(long quotaInBytes,
+                                         long quotaInCounts) {
```
```suggestion
  public static OzoneQuota getOzoneQuota(long quotaInBytes,
      long quotaInCounts) {
```

cxorm commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1706,21 +1706,24 @@ public boolean setOwner(String volume, String owner) throws IOException {
   @Override
-  public void setQuota(String volume, long quota) throws IOException {
-    if (isAclEnabled) {
+  public void setQuota(String volume, long namespaceQuota,
```
Instead of `OzoneManager.java`, this request is processed in `OMVolumeSetQuotaRequest.java` .

We could not update code-snippet here.

captainzmc replied:
Similar to [setOwner](https://github.com/apache/hadoop-ozone/blob/7dac140024214c2189b72fad0566a0252d63e93c/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L1697), there already exists [setQuota](https://github.com/apache/hadoop-ozone/blob/7dac140024214c2189b72fad0566a0252d63e93c/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L1727). If none of this is needed, I think we can separate PR to clean up the useless methods.

cxorm commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/VolumeManagerImpl.java:
```diff
@@ -273,12 +272,14 @@ private void setOwnerCommitToDB(UserVolumeInfo oldOwnerVolumeList,
    */
   @Override
-  public void setQuota(String volume, long quota) throws IOException {
+  public void setQuota(String volume, long namespaceQuota,
```
This part is the same as comments on `OzoneManager.java`

cxorm commented:
Sorry for my careless huge "Add single comment" yesterday.

There are Test-related-part and API-backward-compatibility-part to be reviewed. (If anyone continue reviewing after this comment.)

captainzmc commented:
> Thanks @captainzmc for the work.
> Overall the functionality-part is good and some suggestions is added inline, test-related-part would be reviewed soon.
> 
> Here I have two Questions IMHO.
> 
> (1) Could we let the variable-pair be consistent ?
> I think it would better to use one of (`storagespaceQuota`/ `namespaceQuota`) and (`quotaInBytes`/ `quotaInCounts`).
> If it have other purpose to use both, please let me know.
> 
> (2) Would we add limitation of creating volume and bucket in later patch ?
> I deployed this patch on my machine, and created volume without setting both of quota.
> I found that the storage of volume exceeds my local-storage.
> And I found we could create bucket even if the quota of bucket is -1.

Thanks for @cxorm 's feedback.
1. I will revise the unified naming and other review issues as soon as possible.
2. Currently, quota can only be set, but does not take effect. [This task lists plans and designs](https://issues.apache.org/jira/browse/HDDS-541). I will continue to improve this part in the future.

captainzmc commented:
Thanks for @cxorm 's review. I had revised the review issues. Could you take another look？

captainzmc commented:
Update PR, like HDDS-3991, Ignore protobuf lock files.
Hi @cxorm. Previous issues has been solved. Could you help take another look？

captainzmc commented:
hi @xiaoyuyao @arp7 This is the subtask of HDDS-541 Ozone Quota Support. Could you help review it？

cxorm commented:
> Update PR, like [HDDS-3991](https://issues.apache.org/jira/browse/HDDS-3991), Ignore protobuf lock files.
> Hi @cxorm. Previous issues has been solved. Could you help take another look？

Sorry for late reply, I would take another look.

cxorm commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java:
```diff
@@ -271,10 +271,10 @@ public void testSetVolumeQuota()
     OzoneVolume volume = store.getVolume(volumeName);
-    Assert.assertEquals(100000000L, volume.getQuota());
+    Assert.assertEquals(1024 * 1024 * 1024,
+        volume.getQuotaInBytes());
```
The same as above.
We should verify two quota attributes if the `OzoneQuota` has two attributes.

captainzmc commented on hadoop-ozone/dist/src/main/smoketest/topology/loaddata.robot:
```diff
@@ -25,7 +25,7 @@ Test Timeout        5 minutes
 *** Test Cases ***
 Create a volume, bucket and key
-    ${output} =         Execute          ozone sh volume create topvol1 --quota 100TB
+    ${output} =         Execute          ozone sh volume create topvol1
```
This script is also used in the 0.5->0.6 upgrade. This script was created in the 0.5 environment and the 0.5 version does not support spaceQuota.

captainzmc commented:
> Thanks @captainzmc for working on this patch.
> 
> This review is about test.
> Some suggestions are added inline.

Thanks for @cxorm’s  review. Fixed review issues.

cxorm commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmVolumeArgs.java:
```diff
@@ -62,13 +64,15 @@
   private OmVolumeArgs(String adminName, String ownerName, String volume,
-                       long quotaInBytes, Map<String, String> metadata,
-                       OmOzoneAclMap aclMap, long creationTime,
-                       long modificationTime, long objectID, long updateID) {
+                       long quotaInBytes, long quotaInCounts,
+                       Map<String, String> metadata, OmOzoneAclMap aclMap,
+                       long creationTime, long modificationTime, long objectID,
+                       long updateID) {
```
The same line could be indented by 4 space.

cxorm commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocol/OzoneManagerProtocol.java:
```diff
@@ -88,10 +88,12 @@
    * @throws IOException
    */
-  void setQuota(String volume, long quota) throws IOException;
+  void setQuota(String volume, long quotaInCounts, long quotaInBytes)
+      throws IOException;
```
The same as `ClientProtocol.java` part.
For backward compatibility, I think we should add this method and also keep the original method.

cxorm commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1741,29 +1741,15 @@ public boolean setOwner(String volume, String owner) throws IOException {
-          auditMap, ex));
-      throw ex;
-    }
+  public void setQuota(String volume, long quotaInCounts,
+                       long quotaInBytes) throws IOException {
+    throw new UnsupportedOperationException("OzoneManager does not require " +
+        "this to be implemented. As this requests use a new approach");
```
As mentioned in `OzoneManagerProtocol.java`.

cxorm commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/CreateVolumeHandler.java:
```diff
@@ -40,10 +41,13 @@
+  @Option(names = {"--spaceQuota", "-s"},
+      description = "Quota in bytes of the newly created volume (eg. 1GB)")
+  private String quotaInBytes;
+
   @Option(names = {"--quota", "-q"},
-      description =
-          "Quota of the newly created volume (eg. 1G)")
-  private String quota;
+      description = "Bucket counts of the newly created volume (eg. 5)")
+  private long quotaInCounts = OzoneConsts.QUOTA_COUNT_RESET;
```
Some discussion on naming parameters is updated in design-doc.

We could fix/update here after discussion.

cxorm commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java:
```diff
@@ -258,14 +258,18 @@ public boolean setOwner(String volume, String owner) throws IOException {
   @Override
-  public void setQuota(String volume, long quota) throws IOException {
+  public void setQuota(String volume, long quotaInCounts,
+      long quotaInBytes) throws IOException {
     SetVolumePropertyRequest.Builder req =
         SetVolumePropertyRequest.newBuilder();
-    req.setVolumeName(volume).setQuotaInBytes(quota);
+    req.setVolumeName(volume)
+       .setQuotaInBytes(quotaInBytes)
+       .setQuotaInCounts(quotaInCounts);
```
As mentioned in `OzoneManagerProtocol.java`.

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java:
```diff
@@ -94,13 +98,15 @@
   @SuppressWarnings("parameternumber")
   public OzoneVolume(ConfigurationSource conf, ClientProtocol proxy,
       String name, String admin, String owner, long quotaInBytes,
-      long creationTime, List<OzoneAcl> acls, Map<String, String> metadata) {
+      long quotaInCounts, long creationTime, List<OzoneAcl> acls,
+      Map<String, String> metadata) {
     Preconditions.checkNotNull(proxy, "Client proxy is not set.");
     this.proxy = proxy;
     this.name = name;
     this.admin = admin;
     this.owner = owner;
     this.quotaInBytes = quotaInBytes;
+    this.quotaInCounts = quotaInCounts;
     this.creationTime = Instant.ofEpochMilli(creationTime);
     this.acls = acls;
     this.listCacheSize = HddsClientUtils.getListCacheSize(conf);
```
For backward compatibility, I think we should add new API for new attribute and also keep the original public method.

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java:
```diff
@@ -118,19 +124,19 @@ public OzoneVolume(ConfigurationSource conf, ClientProtocol proxy,
   public OzoneVolume(ConfigurationSource conf, ClientProtocol proxy,
       String name, String admin, String owner, long quotaInBytes,
-      long creationTime, long modificationTime, List<OzoneAcl> acls,
-      Map<String, String> metadata) {
-    this(conf, proxy, name, admin, owner, quotaInBytes,
+      long quotaInCounts, long creationTime, long modificationTime,
+      List<OzoneAcl> acls, Map<String, String> metadata) {
+    this(conf, proxy, name, admin, owner, quotaInBytes, quotaInCounts,
         creationTime, acls, metadata);
     this.modificationTime = Instant.ofEpochMilli(modificationTime);
   }
```
As mentioned above.

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java:
```diff
@@ -141,19 +147,23 @@ public OzoneVolume(ConfigurationSource conf, ClientProtocol proxy,
   @SuppressWarnings("parameternumber")
   public OzoneVolume(ConfigurationSource conf, ClientProtocol proxy,
       String name, String admin, String owner, long quotaInBytes,
-      long creationTime, long modificationTime, List<OzoneAcl> acls) {
-    this(conf, proxy, name, admin, owner, quotaInBytes, creationTime, acls);
+      long quotaInCounts, long creationTime, long modificationTime,
+      List<OzoneAcl> acls) {
+    this(conf, proxy, name, admin, owner, quotaInBytes, quotaInCounts,
+        creationTime, acls);
     this.modificationTime = Instant.ofEpochMilli(modificationTime);
   }
```
As mentioned above.

cxorm commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java:
```diff
@@ -118,19 +124,19 @@ public OzoneVolume(ConfigurationSource conf, ClientProtocol proxy,
   public OzoneVolume(ConfigurationSource conf, ClientProtocol proxy,
       String name, String admin, String owner, long quotaInBytes,
-      long creationTime, List<OzoneAcl> acls) {
-    this(conf, proxy, name, admin, owner, quotaInBytes, creationTime, acls,
-        new HashMap<>());
+      long quotaInCounts, long creationTime, List<OzoneAcl> acls) {
+    this(conf, proxy, name, admin, owner, quotaInBytes, quotaInCounts,
+        creationTime, acls, new HashMap<>());
```
As mentioned above.

captainzmc commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/protocol/ClientProtocol.java:
```diff
@@ -101,10 +100,11 @@ void createVolume(String volumeName, VolumeArgs args)
    * @throws IOException
    */
-  void setVolumeQuota(String volumeName, OzoneQuota quota)
+  void setVolumeQuota(String volumeName, long quotaInBytes, long quotaInCounts)
```
Thanks @cxorm  for the discussion, that's a good question.
The previous feature of Ozone on Set quota is virtual, and this interface won't have any actual effect. This feature is incomplete, so no one will use it before.
One way to think about it is that this is the new API, and subsequent futures will gradually add new interfaces including clrQuota, etc. Users who want to use full functionality of quota need to update the client.
So I think we can think of this part as a new interface without regard to backward compatibility.

cxorm replied:
> The previous feature of Ozone on Set quota is virtual, and this interface won't have any actual effect. This feature is incomplete, so no one will use it before.

I am fine with it.

captainzmc commented:
> Thanks @captainzmc for working on this.
> 
> This review is mainly about API, and related suggestions are added inline.
> 
> Also, there is discussion on naming parameter of CLI in design-doc.

Hi yisheng, I have replied in design-Doc. We've talked about this before. Finally, we refer to the [parameter of HDFS quota](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/HdfsQuotaAdminGuide.html#Administrative_Commands) so as to maximize retention of user habits.

captainzmc commented:
Had fixed issues as HDDS-4129 suggested. CC @ChenSammi.

cxorm commented:
Thank you @captainzmc for updating the PR. 
Could you rebase it with latest master branch ?

This PR looks good to me, +1 (cc @ChenSammi )

ChenSammi commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -53,26 +114,27 @@ public long getSize() {
-    this.unit = Units.UNDEFINED;
+  private OzoneQuota() {
+    this.quotaInCounts = OzoneConsts.QUOTA_RESET;
+    this.quotaInBytes = OzoneConsts.QUOTA_RESET;
```
rawQuotaInBytes initialization is missed.

ChenSammi commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -82,118 +144,130 @@ public OzoneQuota(long size, Units unit) {
    */
   public static String formatQuota(OzoneQuota quota) {
-    return String.valueOf(quota.size) + quota.unit;
+    return String.valueOf(quota.getRawSize())+ quota.getUnit();
```
need space before "+"

ChenSammi commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -82,118 +144,130 @@ public OzoneQuota(long size, Units unit) {
+      long quotaInCounts) {
 
-    if ((quotaString == null) || (quotaString.isEmpty())) {
+    if ((quotaInBytes == null) || (quotaInBytes.isEmpty())) {
```
Use Strings.isNullOrEmpty instead.

ChenSammi commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -82,118 +144,130 @@ public OzoneQuota(long size, Units unit) {
   @Override
   public String toString() {
-    return size + " " + unit;
+    return "Bytes Quota: " + rawQuotaInBytes.toString() + "\n" +
```
Bytes Quota -> Space Bytes Quota
Counts Quota -> Bucket Counts Quota

ChenSammi commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java:
```diff
@@ -204,6 +204,16 @@ public static Versioning getVersioning(boolean versioning) {
   public static final long MAX_QUOTA_IN_BYTES = 1024L * 1024 * TB;
 
+  /**
+   * Quota RESET default is -1, which means it does not take effect.
```
which means it does not take effect -> which means quota is not set

ChenSammi commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/CreateVolumeHandler.java:
```diff
@@ -40,10 +41,13 @@
+      description = "Quota in bytes of the newly created volume (eg. 1GB)")
+  private String quotaInBytes;
+
   @Option(names = {"--quota", "-q"},
```
--quota  -> bucketQuota
-q -> -bq

captainzmc replied:
This command follows the shell of volume (Ozone sh Volume update/create), so the volume in volumeQuota is redundant.

ChenSammi commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/CreateVolumeHandler.java:
```diff
@@ -40,10 +41,13 @@
       description = "Owner of the volume")
   private String ownerName;
 
+  @Option(names = {"--spaceQuota", "-s"},
```
-s -> -sq

captainzmc replied:
The abbreviation of this parameter was previously [confirmed with Yisheng](https://github.com/apache/hadoop-ozone/pull/1233#discussion_r458723395). In general, the parameter abbreviation is a single letter, in order to keep the same with other parameters here sq->s.
Of course, For better understanding, I added a detailed description.

cxorm replied:
I'm fine with the comment with @ChenSammi.
Let's update it.

captainzmc replied:
Agreed, I will fix the naming problem.

ChenSammi commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/UpdateVolumeHandler.java:
```diff
@@ -39,22 +40,33 @@
-      description = "Quota of the volume to set"
-          + "(eg. 1G)")
-  private String quota;
+  @Option(names = {"--spaceQuota", "-s"},
```
same as above

ChenSammi reviewed (commented):
A few minor comments. @captainzmc

captainzmc commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:
```diff
@@ -19,32 +19,93 @@
  * a storage volume.
  */
-public class OzoneQuota {
+public final class OzoneQuota {
```
Subsequent bucket quota is also intended to use this class, so we might need a more generic name here.

captainzmc commented:
Thanks for @ChenSammi 's review, the issues has been fixed.

ChenSammi commented:
LGTM +1.  

Thanks @captainzmc for the contribution and @cxorm for review the patch.
